extra_data/reader.py

```diff
@@ -388,6 +388,10 @@ def train_from_id(
             if file is None:
                 continue
 
+            count = file.get_index(source, '')[1][pos]
+            if not count:
+                continue
+
             for key in self.keys_for_source(source):
                 path = '/CONTROL/{}/{}'.format(source, key.replace('.', '/'))
                 source_data[key] = file.file[path][pos]
```

Comment:
If CONTROL data isn't always there, let's also make sure we do the right thing if it's only partly missing. At present we assume that the position of the train in the index is also the position of the data.

```suggestion
            firsts, counts = file.get_index(source, group)
            first, count = firsts[pos], counts[pos]
            if not count:
                continue

            for key in self.keys_for_source(source):
                path = '/CONTROL/{}/{}'.format(source, key.replace('.', '/'))
                source_data[key] = file.file[path][first]
```

Reply:
My assumption was that the data is either present or entirely missing, but you're right, better be safe!